cpp/src/parquet/page_index.cc

```diff
@@ -184,8 +185,294 @@ class OffsetIndexImpl : public OffsetIndex {
+    /// Simply returns nullptr.
+    return nullptr;
+  }
+
+  void WillNeed(const std::vector<int32_t>& row_group_indices,
+                const std::vector<int32_t>& column_indices,
+                const IndexSelection& index_selection) override {
+    std::vector<::arrow::io::ReadRange> read_ranges;
+    for (int32_t row_group_ordinal : row_group_indices) {
+      auto read_range = PageIndexReader::DeterminePageIndexRangesInRowGroup(
+          *file_metadata_->RowGroup(row_group_ordinal), column_indices);
+      if (index_selection.column_index && read_range.column_index.has_value()) {
+        read_ranges.push_back(*read_range.column_index);
+      } else {
+        // Mark the column index as not requested.
+        read_range.column_index = std::nullopt;
+      }
+      if (index_selection.offset_index && read_range.offset_index.has_value()) {
+        read_ranges.push_back(*read_range.offset_index);
+      } else {
+        // Mark the offset index as not requested.
+        read_range.offset_index = std::nullopt;
+      }
+      index_read_ranges_.emplace(row_group_ordinal, std::move(read_range));
+    }
+    PARQUET_THROW_NOT_OK(input_->WillNeed(read_ranges));
+  }
+
+  void WillNotNeed(const std::vector<int32_t>& row_group_indices) override {
+    for (int32_t row_group_ordinal : row_group_indices) {
+      index_read_ranges_.erase(row_group_ordinal);
+    }
+  }
+
+ private:
+  /// The input stream that can perform random read.
+  ::arrow::io::RandomAccessFile* input_;
+
+  /// The file metadata to get row group metadata.
+  std::shared_ptr<FileMetaData> file_metadata_;
+
+  /// Reader properties used to deserialize thrift object.
+  const ReaderProperties& properties_;
+
+  /// File-level decrypter.
+  std::shared_ptr<InternalFileDecryptor> file_decryptor_;
+
+  /// Coalesced read ranges of page index of row groups that have been suggested by
+  /// WillNeed(). Key is the row group ordinal.
+  std::unordered_map<int32_t, RowGroupIndexReadRange> index_read_ranges_;
+};
+
 }  // namespace
 
+RowGroupIndexReadRange PageIndexReader::DeterminePageIndexRangesInRowGroup(
+    const RowGroupMetaData& row_group_metadata, const std::vector<int32_t>& columns) {
+  int64_t ci_start = std::numeric_limits<int64_t>::max();
+  int64_t oi_start = std::numeric_limits<int64_t>::max();
+  int64_t ci_end = -1;
+  int64_t oi_end = -1;
+
+  auto merge_range = [](const std::optional<IndexLocation>& index_location,
+                        int64_t* start, int64_t* end) {
+    if (index_location.has_value()) {
+      if (index_location->offset < 0 || index_location->length <= 0) {
+        throw ParquetException("Invalid index location: offset ", index_location->offset,
+                               " length ", index_location->length);
+      }
+      *start = std::min(*start, index_location->offset);
+      *end = std::max(*end, index_location->offset + index_location->length);
```

Comment:
Do we want to guard against addition overflow? We try to make the Parquet reader robust against invalid/corrupt data (some fuzzing is done, but it doesn't involve optional features as bloom filters, page index, encryption...).

Reply:
You can use `AddWithOverflow` for that.